src/top/utop.el

```diff
@@ -604,12 +604,10 @@ it is started."
          (dolist (prefix utop-completion-prefixes)
            ;; We need to handle specially prefixes like "List.m" as
            ;; the responses from utop don't include the module prefix.
-           (let ((prefix (if (string-match-p "\\." prefix)
-                             (cadr (split-string prefix "\\."))
-                           prefix)))
-             (when (string-prefix-p prefix argument)
-               (push argument utop-completion)
-               (throw 'done t))))))
+           (when-let* ((pos (string-match  "[^.]*$" prefix))
+                       ((string-prefix-p (substring prefix pos) argument)))
```

Comment:
I don't like much pushing conditional checks without bindings to `when-let`, as this makes a bit harder to understand what the code does.